### PR TITLE
Implement numpy serializer

### PR DESCRIPTION
**Pull request by okuta**

### CODE REVIEW

File: chainer/serializers/npz.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    This is the standard serializer in Chainer. The chain hierarchy is simply
+    mapped to dictionary key.
+
+    Args:
+        dict (dict): The dict that this serializer represents.
+        path (str): The base path.
+
+    """
+    def __init__(self, dict, path=''):
+        self.dict = dict
```
Comment (unnonouno) [inline chainer/serializers/npz.py]: Don't use `dict` as it is reserved.

File: chainer/serializers/npz.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        self.path = path
+
+    def __getitem__(self, key):
+        return DictionarySerializer(self.dict, self.path + key + '/')
```
Comment (unnonouno) [inline chainer/serializers/npz.py]: Do we need to escape the key?

File: chainer/serializers/npz.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    def __call__(self, key, value):
+        ret = value
+        if isinstance(value, cuda.ndarray):
```
Comment (unnonouno) [inline chainer/serializers/npz.py]: I think we don't need this check.

Comment (beam2d) [inline chainer/serializers/npz.py]: This check is required to accept Python scalar values (to_cpu only accepts numpy/cupy arrays).

**Comment by beam2d** [comment]
Fixed. I once think about escaping slashes in the keys, but it made the serialization of optimizers full of escapes, so I do not include the escaping in the current version.

**Comment by unnonouno** [comment]
OK. It looks good. Please fix the document.

**Comment by beam2d** [comment]
Added a document. I also added a compression support.

**Comment by beam2d** [comment]
Added tests without compression.

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**